[SI-527] Remove QValue from MediaRange and MediaType

jedesah opened the pull request:
Not finished, still a few failing tests, but posting to get early feedback

bryce-anderson commented on core/src/main/scala/org/http4s/headers/Accept.scala:
```diff
@@ -1,12 +1,25 @@
 
-object Accept extends HeaderKey.Internal[Accept] with HeaderKey.Recurring
+object Accept extends HeaderKey.Internal[Accept] with HeaderKey.Recurring {
+  implicit def withDefaultQValue(mediaRange: MediaRange): MediaRangeAndQValue =
```
Whats the point of the implicit here? for implicit conversion?

jedesah replied:
Ya, I gather you are not a fan?
On Feb 21, 2016 4:17 PM, "Bryce Anderson" <notifications@github.com> wrote:

> In core/src/main/scala/org/http4s/headers/Accept.scala
> <https://github.com/http4s/http4s/pull/532#discussion_r53576111>:
>
> >  import scalaz.NonEmptyList
> >
> > -object Accept extends HeaderKey.Internal[Accept] with HeaderKey.Recurring
> > +object Accept extends HeaderKey.Internal[Accept] with HeaderKey.Recurring {
> > +  implicit def withDefaultQValue(mediaRange: MediaRange): MediaRangeAndQValue =
>
> Whats the point of the implicit here? for implicit conversion?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/http4s/http4s/pull/532/files#r53576111>.
>

bryce-anderson replied:
I think most people are moving away from implicit conversions. We've made a pattern of using overloaded `apply` functions for other situations similar to this and I would probably have tried that first here (can't say it would have been successful :smile:).

rossabaker replied:
Without this, either all the constants in `MediaType` need to exist on `MediaRangeAndQValue`, or else the conversion is explicit:

```scala
Accept(`text/xml`.withDefaultQValue)
```

bryce-anderson commented:
:+1: I like where this is going. I don't personally like the idea of implicit conversions but am open to hearing a case for the contrary.

rossabaker commented:
Interesting.

Compare this to the current `Charset` hierarchy:
1) `Charset` is analogous to `MediaType`: atoms with no q-value.
2) `CharsetRange` is analogous to `MediaTypeWithQValue`: have a q-value, and can represent the asterisks.
3) `CharsetRange.

rossabaker commented:
Thanks.  I like this better than what's on master.  I'm comparing this to the current `Charset`.

Similar:
1) `Charset` and `MediaType` are consistent: atoms with no q-value.
2) `CharsetRange` and `MediaRange` both introduce the wildcard types.

Different:
1) `Charset` is not a `CharsetRange`, but `MediaType` is a `MediaRange`.
2) `CharsetRange` can be an `Atom` or a `*`, whereas there is no type-level distinction in `MediaRange`.
3) `CharsetRange` introduces the q-value, whereas this is deferred to `MediaRangeWithQValue`.

I don't have a well-formed opinion about which of these models is better.  The main conceptual difference I see between the charset hierarchy and the media-type hierarchy is that the latter has more wildcard types.  But I think whatever we settle on here should be a model for Charset, ContentCoding, and LanguageTag.

jedesah commented:
This is ready to be merged if it passes review. Just added a commit instead of rebasing so you could review the changes more easily, but I will rebase into a single commit if you think it's ready.

@bryce-anderson I removed the implicit conversion and it does not add that much noise (and it's mostly in tests anyway).

@rossabaker Interesting the comparison of `MediaRange` with `Charset`, `ContentCoding` and `LanguageTag`. I think `MediaRange` could still use a little bit of work before being a model for the others. In particular, I am not convinced `MediaType` being a `MediaRange` is necessarily the best approach. However I don't think this needs to happen in this PR as I think this is an improvement in it's own right. But it's probably worth creating another issue that captures your above reflection.

jedesah commented:
Fixes issue #527

rossabaker commented on core/src/test/scala/org/http4s/parser/AcceptHeaderSpec.scala:
```diff
@@ -82,32 +77,30 @@ class AcceptHeaderSpec extends Specification with HeaderParserHelper[Accept] wit
-        `text/html`.withQValue(q(0.7)),
-        `text/html`.withExtensions(Map("level" -> "1"))
+      parse(value) must be_===(Accept(
+        MediaRangeAndQValue(`text/*`,q(0.3)),
```
Is there any reason a `MediaRange` shouldn't have a `.withQValue` that returns a `MediaRangeAndQValue`?  The new structure seems reasonable, but I think I preferred that old syntax.

jedesah replied:
No reason. I will add it.

rossabaker commented on core/src/test/scala/org/http4s/parser/AcceptHeaderSpec.scala:
```diff
@@ -20,61 +19,57 @@ class AcceptHeaderSpec extends Specification with HeaderParserHelper[Accept] wit
       // Just do a single type
-      val accept = Accept(`audio/*`, `video/*`)
-      parse(accept.value) must be_==(accept)
+      val accept = Accept(MediaRangeAndQValue(`audio/*`), MediaRangeAndQValue(`video/*`))
```
Because I'm compiling this in my head, this change is an example of what becomes necessary without that implicit, right?

jedesah replied:
That is correct.

bryce-anderson commented:
Generally I like it!
My only concern is that making the `MediaRangeAndQValue` instances does seem like a bummer. I think the syntax would be made easier with overloaded constructor methods; has that been looked into? If implicit conversions are still a thing I'm not wholesale opposed to them, I just thought they were generally frowned upon these days.

jedesah commented:
@bryce-anderson What do you mean by overloaded constructor methods?

jedesah commented:
I added back `withQValue` since I think it is a good idea irrespective. It helps a little with the noise. It doesn't "solve" the problem where the `QValue` is `1` though. @bryce-anderson I think you are right that implicit conversions not used to add syntax are generally frowned upon these days. I think they can be useful in low risk situations (like this?). Ultimately it comes down to expressivity vs safety, you can't always have both.

I guess an alternative to an implicit conversion would be a default value on `withQValue` so one could do `text/plain`.withQValue(). It strikes me as somewhat odd syntax though.

bryce-anderson commented:
@jedesah, by overloaded constructor methods I mean't overload the `apply` method for the case classes that need a `MediaRangeAndQValue` to also just a `MediaRange`. Of course this doesn't work well in situations where you want to mix and match.

I think this situation is as good a good candidate for an implicit conversion as they come, its just a matter of if we want them at all. Like I said, I'm not opposed but not exactly in favor either.

jedesah commented:
I also don't have a strong opinion either way. I guess I have a slight preference for the implicit conversions but to be fair I have never been bitten by them quite as hard as others seem to have been. Maybe it's because I have been lucky enough to choose appropriately when to use them and when to avoid them.

rossabaker commented:
I'm reluctant to mix DSL concerns into the core model, but this particular implicit makes the model more pleasant and intuitive to use, in my opinion.

bryce-anderson commented:
@jedesah, I think its up to you if you want to bring the implicit conversions back.

jedesah commented:
@bryce-anderson @rossabaker Decided to re-introduce the implicit conversion after all. Let me know if you think this is now good to merge and I can rebase and then hopefully get this merged.

bryce-anderson commented:
:+1: LGTM. I'll merge it this evening so long as Travis gives the OK. If @rossabaker beats me to it, thats fine too. :smile:

bryce-anderson commented:
Thanks @jedesah for the great code!
